maths/modular_inverse.nim

```diff
@@ -0,0 +1,65 @@
+## Modular inverse
+
+import std/[options, math]
+
+
+func euclidHalfIteration(inA, inB: Positive): (Natural, int) =
+  var (a, b) = (inA.Natural, inB.Natural)
+  var (x0, x1) = (1, 0)
+  while b != 0:
+    (x0, x1) = (x1, x0 - (a div b) * x1)
+    (a, b) = (b, a mod b)
+
+  (a, x0)
+
+
+func modularInverse*(inA: int, modulus: Positive): Option[Positive] =
+  ## For given integer `a` and a natural number `modulus` it
+  ## computes the inverse of `a` modulo `modulus`, i.e.
+  ## it finds an integer `0 < inv < modulus` such that
+  ## `(a * inv) mod modulus == 1`.
+  let a = math.floorMod(inA, modulus)
+  if a == 0:
+    return none(Positive)
+  let (gcd, x) = euclidHalfIteration(a, modulus)
+  if gcd == 1:
+    return some(math.floorMod(x, modulus).Positive)
+  none(Positive)
+
+
+when isMainModule:
+  import std/[unittest, sequtils, strformat]
+  suite "modularInverse":
+    const testCases = [
+      (3, 7, 5),
+      (-1, 5, 4),
```

Comment:
```suggestion
      (-1, 5, 4), # Inverse of a negative
      (-7, 5, 2), # Inverse of a negative lower than modulus
      (-7, 4, 1), # Inverse of a negative with non-prime modulus
```
These two test cases felt lacking to me. Why not a randomized test here too?

Reply:
> These two cases have to be tested as well. Why not a randomized test here too?

I agree that writing some _randomized tests_ here is quite straightforward. I am not sure if that would not be an _overkill_ in this situation. Do you really think that feeding _random_ inputs into this function will help us to discover some bug?

Reply:
Just writing a property-based test to check that the obtained inverse makes 1 when multiplied by the initial element is not an overkill.
It is *pedagogical*.

Reply:
I am a bit lost. We have a check if the test case is valid:
https://github.com/TheAlgorithms/Nim/blob/507792656a53ef180f121b490a0ff333ce94ff05/maths/modular_inverse.nim#L50

Reply:
I wanted to test the definition of the inverse on randomized inputs. Pseudocode:
```nim
a = randint(0, modulus-1)
inv = modularInverse(a, modulus)
assert a * inv == 1
```

Reply:
Please have a look at 2e25c2ad1f0dd14daf14dd58de2b300fa9e98253. The added test does not make any assumptions of the input (e.g. `a` can be bigger than `modulus` or negative). In my opinion, it is better, since we test the _entire implementation_.

[off-topic]: I would highly suggest to change the setup of this project in such a way, that a call `nim test randomTestSize=10000` would be possible. The command `nim test` would run only the _deterministic tests_ and would be used for things like generating coverage reports etc.